### PR TITLE
Introduce AbstractOptimizer

### PR DESCRIPTION
**Pull request by vitornesello**

### CODE REVIEW

File: src/MOIinterface.jl [hunk 1]
```diff
@@ -38,191 +25,204 @@ function compute_moi_terms(members::VarMembership)
...
     moi_set = (kind == Binary ? MOI.ZeroOne() : MOI.Integer())
     setkind!(moirecord, MOI.add_constraint(
-        optimizer, MOI.SingleVariable(getindex(moirecord)), moi_set
+        inner, MOI.SingleVariable(getindex(moirecord)), moi_set
     ))
     return
 end
 
-function add_to_optimzer!(optimizer::MOI.AbstractOptimizer, v::Variable)
+function add_to_optimzer!(optimizer::MoiOptimizer, v::Variable)
```
Comment (guimarqu) [inline src/MOIinterface.jl]: add_to_optimizer!

File: src/MOIinterface.jl [hunk 2]
```diff
@@ -38,191 +25,204 @@ function compute_moi_terms(members::VarMembership)
...
-        optimizer, MOI.SingleVariable(getindex(moirecord)),
+        getinner(optimizer), MOI.SingleVariable(getindex(moirecord)),
         MOI.Interval(getlb(cur_data), getub(cur_data))
     )
     setbounds!(moirecord, moi_bounds)
     return
 end
 
-function enforce_var_kind_in_optimizer(optimizer::MOI.AbstractOptimizer,
+function enforce_var_kind_in_optimizer(optimizer::MoiOptimizer,
```
Comment (guimarqu) [inline src/MOIinterface.jl]: enforce_var_kind_in_optimizer!

File: src/MOIinterface.jl [hunk 3]
```diff
@@ -38,191 +25,204 @@ function compute_moi_terms(members::VarMembership)
...
     if (getkind(cur_data) != Binary)
         enforce_bounds_in_optimizer(optimizer, v)
     end
-    MOI.set(optimizer, MOI.VariableName(), moi_index, getname(v))
+    MOI.set(inner, MOI.VariableName(), moi_index, getname(v))
     return
 end
 
-function add_to_optimzer!(optimizer::MOI.AbstractOptimizer,
+function add_to_optimzer!(optimizer::MoiOptimizer,
```
Comment (guimarqu) [inline src/MOIinterface.jl]: add_to_optimizer!

File: src/MOIinterface.jl [hunk 4]
```diff
@@ -38,191 +25,204 @@ function compute_moi_terms(members::VarMembership)
...
                                 constr::Constraint)
     moirecord = getmoirecord(constr)
     @assert getindex(moirecord).value != -1
-    MOI.delete(optimizer, getindex(moirecord))
+    MOI.delete(getinner(optimizer), getindex(moirecord))
     setindex!(moirecord, MoiConstrIndex())
     return
 end
 
+function fill_primal_result(optimizer::MoiOptimizer,
```
Comment (guimarqu) [inline src/MOIinterface.jl]: fill_primal_result!

File: src/MOIinterface.jl [hunk 5]
```diff
@@ -38,191 +25,204 @@ function compute_moi_terms(members::VarMembership)
...
+            end
+        end
+        push!(result.primal_sols, PrimalSolution{S}(pb, sol))
+    end
+    result.primal_bound = getbound(getbestprimalsol(result))
+    @logmsg LogLevel(-2) string("Primal bound is ", getprimalbound(result))
+    return
+end
+
+function fill_dual_result(optimizer::MoiOptimizer,
```
Comment (guimarqu) [inline src/MOIinterface.jl]: fill_dual_result!

File: src/formulation.jl [hunk 6]
```diff
@@ -71,8 +65,8 @@ getuid(f::Formulation) = f.uid
 "Returns the objective function sense of `Formulation` `f`."
 getobjsense(f::Formulation) = f.obj_sense
 
-"Returns the `MOI.Optimizer` of `Formulation` `f`."
-get_optimizer(f::Formulation) = f.moi_optimizer
+"Returns the `AbstractOptimizer` of `Formulation` `f`."
+get_optimizer(f::Formulation) = f.optimizer
```
Comment (guimarqu) [inline src/formulation.jl]: getoptimizer

File: src/old/algpreprocessnode.jl [hunk 7]
```diff
@@ -778,7 +778,7 @@
 #     master = alg.extended_problem.master_problem
 #     infeas_cols = find_infeasible_columns(master, alg.preprocessed_vars)
 
-#     update_moi_optimizer(
+#     update_optimizer(
```
Comment (guimarqu) [inline src/old/algpreprocessnode.jl]: update_optimizer!

File: src/optimizerwrappers.jl [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    sync_solver(form, get_optimizer(form))
+    @logmsg LogLevel(-3) "MOI formulation after synch: "
+    @logmsg LogLevel(-3) get_optimizer(form)
+    call_moi_optimize_with_silence(form.optimizer)
+    status = MOI.get(form.optimizer.inner, MOI.TerminationStatus())
+    @logmsg LogLevel(-2) string("Optimization finished with status: ", status)
+    return retrieve_result(form, optimizer)
+end
+
+function sync_solver(f::Formulation, optimizer::MoiOptimizer)
```
Comment (guimarqu) [inline src/optimizerwrappers.jl]: sync_solver!(optimizer::MoiOptimizer, f::Formulation)

File: src/problem.jl [hunk 9]
```diff
@@ -42,8 +42,8 @@ set_re_formulation!(m::Problem, r::Reformulation) = m.re_formulation = r
 get_original_formulation(m::Problem) = m.original_formulation
 get_re_formulation(m::Problem) = m.re_formulation
 
-function initialize_moi_optimizer(prob::Problem)
-    initialize_moi_optimizer(
+function initialize_optimizer(prob::Problem)
```
Comment (guimarqu) [inline src/problem.jl]: initialize = init ?

Comment (vitornesello) [reply src/problem.jl]: this will maybe change, so we see later

File: src/old/algpreprocessnode.jl (hunk 7 above)
Comment (vitornesello) [reply src/old/algpreprocessnode.jl]: this is in old/

**Pull request merged**